File: packages/react-native-performance/ios/ReactNativePerformance/ReactNativePerformanceModule.m
#import <React/RCTBridgeModule.h>

@interface RCT_EXTERN_REMAP_MODULE(Performance, ReactNativePerformanceModule, NSObject)

RCT_EXTERN_METHOD(getNativeStartupTimestamp:(RCTPromiseResolveBlock)resolve reject:(RCTPromiseRejectBlock)reject)
RCT_EXTERN_METHOD(getNativeTime:(RCTPromiseResolveBlock)resolve reject:(RCTPromiseRejectBlock)reject)
RCT_EXTERN_METHOD(getNativeUuid:(RCTPromiseResolveBlock)resolve reject:(RCTPromiseRejectBlock)reject)

@end

File: packages/react-native-performance/ios/ReactNativePerformance/ReactNativePerformanceModule.swift
import Foundation

@objc(ReactNativePerformanceModule)
class ReactNativePerformanceModule: NSObject {
    @objc func getNativeStartupTimestamp(_ resolve: RCTPromiseResolveBlock, reject: RCTPromiseRejectBlock) {
        let startupTimestamp = ReactNativePerformance.startupTimestamp()
        if (startupTimestamp < 0.0) {
            let message = "[ReactNativePerformance onAppStarted] was never called by the native portion of the app."
            reject(nil, message, PerformanceError.illegalState(message))
            return
        }

        resolve(String(startupTimestamp))
    }

    @objc func getNativeTime(_ resolve: RCTPromiseResolveBlock, reject: RCTPromiseRejectBlock) {
        let timestamps = [
            "timeSinceEpochMillis": String(Timestamp.nowMillis()),
            "uptimeMillis": String(Timestamp.systemUptimeMillis())
        ]

        resolve(timestamps)
    }

    @objc func getNativeUuid(_ resolve: RCTPromiseResolveBlock, reject: RCTPromiseRejectBlock) {
        resolve(NSUUID().uuidString)
    }

    @objc static func requiresMainQueueSetup() -> Bool {
        return true
    }
}
